SpatialGDK/Source/SpatialGDK/Public/Schema/ComponentPresence.h

```diff
@@ -0,0 +1,121 @@
+
+	ComponentPresence(TArray<Worker_ComponentId> InActorComponentList)
+		: ComponentList(InActorComponentList)
+	{}
+
+	ComponentPresence(const Worker_ComponentData& Data)
+	{
+		Schema_Object* ComponentObject = Schema_GetComponentDataFields(Data.schema_type);
+		CopyListFromComponentObject(ComponentObject);
+	}
+
+	Worker_ComponentData CreateComponentPresenceData()
+	{
+		return CreateComponentPresenceData(ComponentList);
+	}
+
+	static Worker_ComponentData CreateComponentPresenceData(const TArray<Worker_ComponentId>& ComponentList)
+	{
+		Worker_ComponentData Data = {};
+		Data.component_id = ComponentId;
+		Data.schema_type = Schema_CreateComponentData();
+		Schema_Object* ComponentObject = Schema_GetComponentDataFields(Data.schema_type);
+
+		for (int32 i = 0; i < ComponentList.Num(); ++i)
+		{
+			Schema_AddUint32(ComponentObject, SpatialConstants::COMPONENT_PRESENCE_COMPONENT_LIST_ID, ComponentList[i]);
+		}
+
+		return Data;
+	}
+
+	Worker_ComponentUpdate CreateComponentPresenceUpdate()
+	{
+		Worker_ComponentUpdate Update = {};
+		Update.component_id = ComponentId;
+		Update.schema_type = Schema_CreateComponentUpdate();
+		Schema_Object* ComponentObject = Schema_GetComponentUpdateFields(Update.schema_type);
+
+		Schema_AddUint32List(ComponentObject, SpatialConstants::COMPONENT_PRESENCE_COMPONENT_LIST_ID,
+			ComponentList.GetData(), ComponentList.Num());
+
+		return Update;
+	}
+
+	void ApplyComponentUpdate(const Worker_ComponentUpdate& Update)
+	{
+		Schema_Object* ComponentObject = Schema_GetComponentUpdateFields(Update.schema_type);
+		CopyListFromComponentObject(ComponentObject);
+	}
+
+	void CopyListFromComponentObject(Schema_Object* ComponentObject)
+	{
+		ComponentList.SetNum(Schema_GetUint32Count(ComponentObject, SpatialConstants::COMPONENT_PRESENCE_COMPONENT_LIST_ID), true);
+		Schema_GetUint32List(ComponentObject, SpatialConstants::COMPONENT_PRESENCE_COMPONENT_LIST_ID, ComponentList.GetData());
+	}
+
+	bool AddComponentDataIds(const TArray<FWorkerComponentData>& ComponentDatas)
+	{
+		TArray<Worker_ComponentId> ComponentIds;
+		ComponentIds.SetNum(ComponentDatas.Num());
+		for (const FWorkerComponentData& ComponentData : ComponentDatas)
+		{
+			ComponentIds.Add(ComponentData.component_id);
+		}
+
+		return AddComponentIds(ComponentIds);
+	}
+
+	bool AddComponentIds(const TArray<Worker_ComponentId>& ComponentsToAdd)
+	{
+		const int32 OldPresentComponentCount = ComponentList.Num();
+
+		for (auto& NewComponentId : ComponentsToAdd)
+		{
+			ComponentList.AddUnique(NewComponentId);
+		}
+
+		return OldPresentComponentCount != ComponentList.Num();
+	}
+
+	bool RemoveComponentIds(const TArray<Worker_ComponentId>& ComponentsToRemove)
+	{
+		const int32 OldPresentComponentCount = ComponentList.Num();
+
+		for (auto RequiredComponentIt = ComponentList.CreateIterator(); RequiredComponentIt; RequiredComponentIt++)
+		{
+			if (ComponentsToRemove.Contains(*RequiredComponentIt))
+			{
+				RequiredComponentIt.RemoveCurrent();
```

Comment:
Do unreal iterators let you do this? Is there not a predicated remove function or something else we can do this with?

Reply:
Yep you're right, it invalidates